proposals/exception-handling/Exceptions.md

```diff
@@ -288,28 +288,103 @@ end
 If `call $foo` throws, searching for a catching block first finds `delegate`,
 and because it delegates exception handling to catching instructions associated
 with `$l1`, it will be next checked by the outer `catch` and then `catch_all`
-instructions. When the specified label within a `delegate` instruction does not
-correspond to a `try` instruction, it is a validation failure.
+instructions.
+
+`delegate` can also target `catch`-less `try`s or non-`try` block constructs
+like `block`s or `loop`s, in which case the delegated exception is assumed to
+propagate to the outer scope and will be caught by the next matching
+try-catches, or rethrown to the caller if there is no outer try block. In the
+examples, catches are annotated with `($label_name)` to clarify which `try` it
+belongs to for clarification; it is not the official syntax.
+```
+try $l0
+  block $l1
+    try
+      call $foo
+    delegate $l1  ;; delegates to 'catch ($l0)'
+  end
+catch ($l0)
+end
+```
 
-Note that the example below is a validation failure:
+`delegate` can only target `catch`/`catch_all`s that are below the current
+instruction, i.e., it can only delegate downwards. It also targets all
+`catch`/`catch_all`s of a `try` as a whole and does not target individual
+`catch`/`catch_all` within a `try`. Here is another example:
 ```
-try $l1
-catch
+try $l0
+  try $l1
+  catch ($l1)
+    try
+      call $foo
+    delegate $l1  ;; delegates to 'catch ($l0)'
+  catch_all
+    ...
+  end
+catch ($l0)
+```
+
+Here the `delegate` is targeting `catch ($l1)`, which exists before the
+`delegate`. So in case an exception occurs, it propagates out and ends up
+targetting `catch ($l0)`, if the catch has a matching tag. If not, it will
+propagate further out. Even if the `catch_all` is below the `delegate`,
+`delegate` targets catches of a `try` as a whole and does not target an
+individual `catch`/`catch_all`, so it doesn't apply.
+
+If `delegate` targets the implicit function body block, then in effect it delegates the
+exception to the caller of the current function. For example:
+```
+(func $test
   try
-    call $foo
-  delegate $l1  ;; (= delegate 0)
-catch_all
-  ...
-end
+    try
+      call $foo
+    delegate 1  ;; delegates to the caller
+  catch
+    ...
+  catch_all
+    ...
+  end
+)
+```
+In case `foo` throws, `delegate 1` here delegates the exception handling to the
+caller, i.e., the exception escapes the current function. If the immediate is
+greater than the depth of the outermost block + 1, it is a validation failure.
```

Comment:
Should this be "greater than or equal to"? (or "depth of the outermost block" without +1?)

Reply:
Hmm, yeah, "the depth of the outermost block" is confusing. I tried to rephrase it to
> If the immediate is greater than the number of block nesting including the implicit function-level block, it is a validation failure.

Do you think it's better?

Reply:
I think there's still some potential for confusion if someone interprets "number of block nesting" as counting from 1 and not 0. (In the example right above here you might count nestings of `try` and function body as 2, but immediates >=2 are not allowed here) Perhaps it could explicitly say it's counting from zero?

Reply:
Sorry, yeah, it should change to "greater than equal to". Intuitively the level of nesting is a number > 0 when there's any, and that's what I meant too. I changed it to
> If the immediate is greater than or equal to the number of block nesting including the implicit function-level block, it is a validation failure. In this example, any number equal to or greater than 2 is now allowed.

Reply:
Oh... now allowed -> not allowed...